src/randomness/prng.rs

```diff
@@ -0,0 +1,297 @@
+        self.random_bytes(&mut buf);
+        let random_value = u64::from_le_bytes(buf);
+
+        if min == 0 && max == u64::MAX {
+            random_value
+        } else {
+            min + (random_value % (max - min + 1))
+        }
+    }
+
+    pub fn uniform_int_vec(&mut self, min: u64, max: u64, n_elements: usize) -> Vec<u64> {
+        assert!(min <= max, "Invalid interval");
+        let mut return_vec = Vec::with_capacity(n_elements);
+        for _ in 0..n_elements {
+            return_vec.push(self.uniform_int(min, max));
+        }
+        return_vec
+    }
+
+    pub fn uniform_distinct_int_vec(&mut self, min: u64, max: u64, n_elements: usize) -> Vec<u64> {
+        assert!(min <= max, "Invalid interval");
+        let n_elements_max = if max == min {
+            0
+        } else {
+            ((max - min - 1) / 2 + 1) as usize
+        };
+        assert!(
+            n_elements <= n_elements_max,
+            "Number of elements must be less than or equal to half the number of elements in the interval"
+        );
+        let mut return_vec = Vec::with_capacity(n_elements);
+        let mut current_set = HashSet::new();
+        while current_set.len() < n_elements {
+            let value = self.uniform_int(min, max);
+            if current_set.insert(value) {
+                return_vec.push(value);
+            }
+        }
+        return_vec
+    }
+
+    // XXX : dumb implementation. should reduce calls to uniform_int_vec
+    pub fn uniform_bool_vec(&mut self, n_elements: usize) -> Vec<bool> {
+        let bits = self.uniform_int_vec(0, 1, n_elements);
+        bits.into_iter().map(|bit| bit != 0).collect()
+    }
+
+    pub fn uniform_bigint<B: BigInteger>(&mut self, min: B, max: B) -> B {
+        assert!(min <= max, "Invalid interval");
+        let mut range = max.clone();
+        range.sub_with_borrow(&min);
+        let mut random_value: B;
+
+        let num_bits = range.num_bits() as usize;
+
+        loop {
+            let mut bytes = vec![0u8; B::NUM_LIMBS * 8]; // B::NUM_LIMBS * sizeof(u64)
+            self.random_bytes(&mut bytes);
+
+            // Mask unnecessary bits
+            let full_bytes = num_bits / 8;
+            let remaining_bits = num_bits % 8;
+
+            if remaining_bits != 0 {
+                bytes[full_bytes] &= (1 << remaining_bits) - 1;
+            }
+            for i in (full_bytes + 1)..bytes.len() {
+                bytes[i] = 0;
+            }
+
+            // modify bytes to bool array
+            let mut bits = Vec::with_capacity(B::NUM_LIMBS * 64);
+            for byte in &bytes {
+                for bit in 0..8 {
+                    bits.push((byte >> bit) & 1 == 1);
+                }
+            }
+
+            // then make it to BigInteger
+            random_value = B::from_bits_le(&bits);
+            if random_value <= range {
+                break;
+            }
+        }
+        random_value.add_with_carry(&min);
```

Comment:
Is this correct? The `range` is already `max` so addming `min` takes it out of range no?

Reply:
There's `        range.sub_with_borrow(&min);`